### PR TITLE
constraint gen draft

### PR DESCRIPTION
**Pull request by mlutze**
related to #6448

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    case Expr.Binary(sop, exp1, exp2, tpe, loc) => ???
+    case Expr.IfThenElse(exp1, exp2, exp3, loc) =>
+      val (tpe1, eff1) = visitExp(exp1, renv, lenv)
+      val (tpe2, eff2) = visitExp(exp2, renv, lenv)
+      val (tpe3, eff3) = visitExp(exp3, renv, lenv)
+      unifyTypeM(tpe1, Type.Bool, renv, lenv, exp1.loc)
+      unifyTypeM(tpe2, tpe3, renv, lenv, loc)
+      val resTpe = tpe3
+      val resEff = Type.mkUnion(eff1, eff2, eff3, loc)
+      (resTpe, resEff)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:51]: This seems reasonable, no?

**Comment by mlutze** [comment]
@magnus-madsen Do you want this cleaned before merging or do we just press on?

**Comment by mlutze** [comment]
Everything is there except AE stuff, since there were some rules that need to be refactored to be more constraint-gen friendly.

**Comment by magnus-madsen** [comment]
Let us clean it, but you can fork of this branch into the solver, if you want.

**Comment by magnus-madsen** [comment]
What about some print flags to see the constraints? And some timing measurements?

Does it run?

**Comment by mlutze** [comment]
replaced by #6498

**Pull request closed without merging**